### DISCUSSION TITLE
Automatically append query string parameter to all links?

### DISCUSSION THREAD

**Discussion by BrianVallelunga** [TanStack/router#2703] [Q&A]
I want to automatically modify the `toLocation` of all links when clicked. I'd like to either add a search parameter or modify the path in some way.

I've tried playing with `router.subscribe` but it seems to be read-only. Is there a way to add some middleware that can accomplish this? I saw references to middleware, but none on creating it myself.

For example, assuming a user clicks a link to `https://mysite.com/page` then I would like to modify the URL and send the user to `https://mysite.com/page?x=somevalue` so that is what is shown in the browser.

Thanks

**Comment by CanRau** [comment]
An option would be to wrap `<Link>` and add your custom logic in there, here's the [docs](https://tanstack.com/router/latest/docs/framework/react/guide/custom-link)

**Reply by BrianVallelunga** [reply]
That's certainly an option, but doesn't help when I programmatically navigate someone based on some other interaction such as a form redirect.

**Reply by CanRau** [reply]
You could also wrap the [`useNavigate`](https://tanstack.com/router/latest/docs/framework/react/api/router/useNavigateHook) hook and add your logic

But yea some kind of middleware might be interesting, tho until then 😉

**Comment by schiller-manuel** [comment]
this is what search middlewares are for:

https://tanstack.com/router/v1/docs/framework/react/guide/search-params#transforming-search-with-search-middlewares
